File: go/204.go
/*

https://leetcode.com/problems/count-primes/

Count the number of prime numbers less than a non-negative number, n.

Example:

Input: 10
Output: 4
Explanation: There are 4 prime numbers less than 10, they are 2, 3, 5, 7.

*/

package main


import "fmt"


func countTrue(nums []bool) int {
    res := 0
    for i := 0; i < len(nums); i++ {
        if nums[i] {
            res++
        }
    }
    return res
}


// less than n
func countPrimes(n int) int {
    if n < 2 {
        return 0
    }
    prime := make([]bool, n)
    for i := 0; i < n; i++ {
        prime[i] = true
    }
    prime[0] = false
    prime[1] = false
    for i := 2; i * 2 < n; i++ {
        for j := 2; i * j < n; j++ {
            prime[i * j] = false
        }
    }
    return countTrue(prime)
}


func main() {
    fmt.Println(countPrimes(50))
}

File: python/204.py
"""
https://leetcode.com/problems/count-primes/

Count the number of prime numbers less than a non-negative number, n.

Example:

Input: 10
Output: 4
Explanation: There are 4 prime numbers less than 10, they are 2, 3, 5, 7.

"""


class Solution:
    def countPrimes(self, n: int) -> int:
        if n < 2:
            return 0
        primes = [True] * n
        primes[0] = primes[1] = False
        i = 2
        while i ** 2 < n:
            j = 2
            while i * j < n:
                primes[i * j] = False
                j += 1
            i += 1
        return primes.count(True)



# 这个是效率最高的
class Solution:
    def countPrimes(self, n: int) -> int:
        if n < 2:
            return 0

        primes = [True] * n
        primes[0] = primes[1] = False

        for i in range(2, n):
            k = i ** 2
            if k >= n:
                break
            if not primes[k]:
                continue
            for j in range(k, n, i):
                primes[j] = False

        return primes.count(True)


if __name__ == '__main__':
    for i in range(2, 100):
        print(i, Solution().countPrimes(i))

